# Model Training and Evaluation

In [1]:
# Importing required libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
# Loading CSV file
pca_df = pd.read_csv('final_pca_data.csv')

In [3]:
# Checking information
pca_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 21 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Song    262 non-null    object 
 1   Artist  262 non-null    object 
 2   Label   262 non-null    int64  
 3   PC1     262 non-null    float64
 4   PC2     262 non-null    float64
 5   PC3     262 non-null    float64
 6   PC4     262 non-null    float64
 7   PC5     262 non-null    float64
 8   PC6     262 non-null    float64
 9   PC7     262 non-null    float64
 10  PC8     262 non-null    float64
 11  PC9     262 non-null    float64
 12  PC10    262 non-null    float64
 13  PC11    262 non-null    float64
 14  PC12    262 non-null    float64
 15  PC13    262 non-null    float64
 16  PC14    262 non-null    float64
 17  PC15    262 non-null    float64
 18  PC16    262 non-null    float64
 19  PC17    262 non-null    float64
 20  PC18    262 non-null    float64
dtypes: float64(18), int64(1), object(2)
mem

In [4]:
pca_df.head()

Song                                     Artist  Label         PC1  \
0     Gun Smoke                                  21 Savage      0  217.778072   
1  Bank Account                                  21 Savage      1    8.997858   
2       Monster                                  21 Savage      0   89.551273   
3             X  21 Savage & Metro Boomin Featuring Future      1   50.251343   
4    Youngblood                        5 Seconds Of Summer      1 -175.611672   

         PC2        PC3         PC4         PC5         PC6        PC7  ...  \
0   5.178675  59.523433  -57.545132  109.693361   84.933098 -23.822052  ...   
1   9.232807 -19.107245   -9.207286  -96.725596  113.682720 -41.158857  ...   
2  55.975653 -33.644991  -50.167870  -84.392725   15.009782  78.003824  ...   
3  65.280098 -23.255989   24.892724 -117.853389  -19.419696  61.331924  ...   
4  11.165726  35.649720 -114.647831  -26.336128  -26.758468 -22.840631  ...   

         PC9       PC10       PC11       PC12       PC13       PC14  \
0  88.893056  31.988010  72.013385  22.629417   2.987139  -8.473347   
1 -60.116612   0.894825  23.102465 -54.266331  -2.742280   3.641246   
2  27.299177  45.813469 -32.509676  -0.587473  52.824897 -33.879289   
3 -21.430970 -30.800034 -33.859842 -17.930856  29.336462  43.951412   
4   2.291090 -25.236162 -19.448258  49.935816  46.808053  49.715183   

        PC15       PC16       PC17       PC18  
0 -16.617323  56.510582 -10.716302  -9.978425  
1   2.238065  -9.331464 -54.001478  -7.551556  
2 -57.405587 -39.192014  22.341323   7.381429  
3 -28.697984 -13.807712  15.601290   4.307003  
4  -0.162495  14.693898 -43.389356  25.287289  

[5 rows x 21 columns]

In [5]:
pca_df['Label'].unique()

array([0, 1], dtype=int64)

In [6]:
pca_df.describe().T

count          mean         std         min        25%       50%  \
Label  262.0  8.015267e-01    0.399614    0.000000   1.000000  1.000000   
PC1    262.0 -3.471354e-15  120.195200 -233.663846 -95.833996  1.925126   
PC2    262.0  5.207031e-15   81.972481 -242.007634 -55.195026  3.243480   
PC3    262.0 -9.763183e-16   75.864270 -171.700002 -60.281085 -5.640906   
PC4    262.0 -6.508788e-15   69.532320 -145.569132 -50.167870 -6.402239   
PC5    262.0 -2.169596e-16   67.333299 -206.234936 -44.664861 -2.101419   
PC6    262.0 -4.339192e-15   62.573678 -160.800846 -37.762620 -1.688822   
PC7    262.0  5.532470e-15   57.711407 -141.642616 -39.479009 -2.436777   
PC8    262.0 -4.013753e-15   51.520901 -130.304985 -35.938858  1.757923   
PC9    262.0 -1.952637e-15   48.767904 -136.722240 -30.960038  1.783414   
PC10   262.0 -8.352945e-15   47.463670 -130.938083 -33.848052  1.627451   
PC11   262.0  9.763183e-16   46.518659 -117.870241 -32.341498  0.061065   
PC12   262.0  1.193278e-15   44.972919 -130.231482 -32.464031 -3.736887   
PC13   262.0  2.006876e-15   43.305991 -118.229481 -31.755482  2.362546   
PC14   262.0  5.640950e-15   41.900726 -141.248223 -27.319035  0.772348   
PC15   262.0  6.508788e-16   40.798262 -124.311314 -26.800574 -0.823052   
PC16   262.0 -1.952637e-15   39.369092  -99.928810 -22.691661 -1.023220   
PC17   262.0 -2.169596e-15   38.537971 -100.577945 -26.661236  3.193356   
PC18   262.0 -6.508788e-16   37.837572 -104.382813 -23.550259 -1.154189   

             75%         max  
Label   1.000000    1.000000  
PC1    88.032477  265.277170  
PC2    55.173139  266.530735  
PC3    58.775718  202.817086  
PC4    45.486297  160.839968  
PC5    37.953558  216.265976  
PC6    34.511352  201.303075  
PC7    47.171802  151.002950  
PC8    31.528733  167.053317  
PC9    33.004226  137.727262  
PC10   32.943585  138.721963  
PC11   31.982694  162.147782  
PC12   31.176987  140.387365  
PC13   29.417161  118.567772  
PC14   29.215588  143.850503  
PC15   29.184712  126.885705  
PC16   24.675380  105.999214  
PC17   28.846673  100.879399  
PC18   22.258158  108.558955

In [7]:
# Creating a copy of dataframe pca_df
df = pca_df

In [8]:
df.head()

Song                                     Artist  Label         PC1  \
0     Gun Smoke                                  21 Savage      0  217.778072   
1  Bank Account                                  21 Savage      1    8.997858   
2       Monster                                  21 Savage      0   89.551273   
3             X  21 Savage & Metro Boomin Featuring Future      1   50.251343   
4    Youngblood                        5 Seconds Of Summer      1 -175.611672   

         PC2        PC3         PC4         PC5         PC6        PC7  ...  \
0   5.178675  59.523433  -57.545132  109.693361   84.933098 -23.822052  ...   
1   9.232807 -19.107245   -9.207286  -96.725596  113.682720 -41.158857  ...   
2  55.975653 -33.644991  -50.167870  -84.392725   15.009782  78.003824  ...   
3  65.280098 -23.255989   24.892724 -117.853389  -19.419696  61.331924  ...   
4  11.165726  35.649720 -114.647831  -26.336128  -26.758468 -22.840631  ...   

         PC9       PC10       PC11       PC12       PC13       PC14  \
0  88.893056  31.988010  72.013385  22.629417   2.987139  -8.473347   
1 -60.116612   0.894825  23.102465 -54.266331  -2.742280   3.641246   
2  27.299177  45.813469 -32.509676  -0.587473  52.824897 -33.879289   
3 -21.430970 -30.800034 -33.859842 -17.930856  29.336462  43.951412   
4   2.291090 -25.236162 -19.448258  49.935816  46.808053  49.715183   

        PC15       PC16       PC17       PC18  
0 -16.617323  56.510582 -10.716302  -9.978425  
1   2.238065  -9.331464 -54.001478  -7.551556  
2 -57.405587 -39.192014  22.341323   7.381429  
3 -28.697984 -13.807712  15.601290   4.307003  
4  -0.162495  14.693898 -43.389356  25.287289  

[5 rows x 21 columns]

Preforming one-hot encoding for song and artist columns to use them during modeling

In [9]:
# Performing one-hot encoding on the 'Song' column
Song_encoded = pd.get_dummies(df['Song'], prefix='Song')

# Concatenating the encoded columns with the DataFrame
df = pd.concat([df, Song_encoded], axis=1)

# Drop the original 'Song' column
df.drop('Song', axis=1, inplace=True)

In [10]:
# Performing one-hot encoding on the 'Artist' column
Artist_encoded = pd.get_dummies(df['Artist'], prefix='Artist')

# Concatenating the encoded columns with the DataFrame
df = pd.concat([df, Artist_encoded], axis=1)

# Drop the original 'Artist' column
df.drop('Artist', axis=1, inplace=True)

In [11]:
# CHecking shape of the dataframe after encoding
df.shape

(262, 488)

Model Training and Evaluation:

Logistic Regression:

In [12]:
# Importing essential libraries for Logistic Regression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Step 1: Data Preparation
X = df.drop('Label', axis=1)
y = df['Label']

# Step 2: Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)

# Step 3: Training the model
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

# Step 4: Evaluating the model
y_pred = logistic_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.7904761904761904
Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        21
           1       0.80      0.99      0.88        84

    accuracy                           0.79       105
   macro avg       0.40      0.49      0.44       105
weighted avg       0.64      0.79      0.71       105

Confusion Matrix:
[[ 0 21]
 [ 1 83]]


C:\Users\Basan\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Model Accuracy:-
The accuracy is approximately 79.04% meaning the model correctly predicts
the target variable for about 79.04% of instances in the test set.

Model Evaluation:-
For Unpopular:
Precision: 0 (0% of the instances predicted as class 0 were actually class 0)
Recall: 0 (0% of the actual class 0 instances were predicted as class 0)
F1-score: 0 (harmonic mean of precision and recall)
For Popular:
Precision: 0.8 (80% of the instances predicted as class 1 were actually class 1)
Recall: 0.99 (99% of the actual class 1 instances were predicted as class 1)
F1-score: 0.88 (High F1-score indicates better performance in predicting class 1)

In [13]:
# Making predictions on training and testing data
y_train_predict = logistic_model.predict(X_train)
y_test_predict = logistic_model.predict(X_test)

# Calculate accuracies
train_accuracy = accuracy_score(y_train, y_train_predict)
test_accuracy = accuracy_score(y_test, y_test_predict)

print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)

Training Accuracy: 0.8789808917197452
Testing Accuracy: 0.7904761904761904


Naive Bayes:

In [14]:
# Importing required library for Naive Bayes
from sklearn.naive_bayes import GaussianNB

# Step 1: Data Preparation
X = df.drop('Label', axis=1)
y = df['Label']

# Step 2: Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)

# Step 3: Training the model
naive_bayes_model = GaussianNB()
naive_bayes_model.fit(X_train, y_train)

# Step 4: Evaluating the model
y_pred = naive_bayes_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.20610687022900764
Classification Report:
              precision    recall  f1-score   support

           0       0.17      0.70      0.27        27
           1       0.50      0.08      0.13       104

    accuracy                           0.21       131
   macro avg       0.33      0.39      0.20       131
weighted avg       0.43      0.21      0.16       131

Confusion Matrix:
[[19  8]
 [96  8]]


In [15]:
# Making predictions on training and testing data
y_train_predict = naive_bayes_model.predict(X_train)
y_test_predict = naive_bayes_model.predict(X_test)

# Calculate accuracies
train_accuracy = accuracy_score(y_train, y_train_predict)
test_accuracy = accuracy_score(y_test, y_test_predict)

print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)

Training Accuracy: 1.0
Testing Accuracy: 0.20610687022900764


The training accuracy of 1.0 means that your Naive Bayes model has perfectly learned the patterns in the training data. But, testing accuracy of 0.20 is quite low which suggests that the model is not performing well on unseen data. The low testing accuracy indicates that the model is not generalizing well and is not able to make accurate predictions on new data points.

Decision Trees:

In [19]:
# Importing reqired library for Decision Trees
from sklearn.tree import DecisionTreeClassifier

# Step 1: Data Preparation
X = df.drop('Label', axis=1)
y = df['Label']

# Step 2: Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.41, random_state=42)

# Step 3: Training the model
decision_tree_model = DecisionTreeClassifier()
decision_tree_model.fit(X_train, y_train)

# Step 4: Evaluating the model
y_pred = decision_tree_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.75
Classification Report:
              precision    recall  f1-score   support

           0       0.22      0.09      0.13        22
           1       0.80      0.92      0.85        86

    accuracy                           0.75       108
   macro avg       0.51      0.50      0.49       108
weighted avg       0.68      0.75      0.71       108

Confusion Matrix:
[[ 2 20]
 [ 7 79]]


In [20]:
# Making predictions on training and testing data
y_train_predict = decision_tree_model.predict(X_train)
y_test_predict = decision_tree_model.predict(X_test)

# Calculate accuracies
train_accuracy = accuracy_score(y_train, y_train_predict)
test_accuracy = accuracy_score(y_test, y_test_predict)

print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)

Training Accuracy: 1.0
Testing Accuracy: 0.75


The training accuracy of 1.0 indicates that the decision tree model has perfectly fit the training data, correctly predicting all labels. However, with a testing accuracy of 0.75, the model seems to generalize better to unseen data compared to the Naive Bayes model. This suggests that the decision tree model is more robust and is better for making accurate predictions on new data. 

SVM:

In [37]:
# Importing essential libraries for Support Vector Machine (SVM)
from sklearn.svm import SVC

# Step 1: Data Preparation
X = df.drop('Label', axis=1)
y = df['Label']

# Step 2: Splitting the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.31, random_state=42)

# Step 3: Training the model
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)

# Step 4: Evaluating the model
y_pred = svm_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print("Classification Report:")
print(classification_report(y_test, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.7926829268292683
Classification Report:
              precision    recall  f1-score   support

           0       0.57      0.22      0.32        18
           1       0.81      0.95      0.88        64

    accuracy                           0.79        82
   macro avg       0.69      0.59      0.60        82
weighted avg       0.76      0.79      0.76        82

Confusion Matrix:
[[ 4 14]
 [ 3 61]]


In [38]:
# Making predictions on training and testing data
y_train_predict = svm_model.predict(X_train)
y_test_predict = svm_model.predict(X_test)

# Calculate accuracies
train_accuracy = accuracy_score(y_train, y_train_predict)
test_accuracy = accuracy_score(y_test, y_test_predict)

print("Training Accuracy:", train_accuracy)
print("Testing Accuracy:", test_accuracy)

Training Accuracy: 1.0
Testing Accuracy: 0.7926829268292683


The training accuracy of 1.0 indicates that the SVM model has perfectly fit the training data, correctly predicting all labels. However, with a testing accuracy of 0.79, the model seems to generalize better to unseen data compared to the Naive Bayes model and Decision Trees model. This suggests that the decision tree model is more robust and is better for making accurate predictions on new data.